services/imageproxy/imageproxy.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/mattermost/mattermost-server/services/httpservice"
+)
+
+// An ImageProxy is the public interface for Mattermost's image proxy. An instance of ImageProxy should be created
+// using MakeImageProxy which requires a configService and an HTTPService provided by the server.
+type ImageProxy struct {
+	ConfigService    configservice.ConfigService
+	configListenerId string
+
+	HTTPService httpservice.HTTPService
+
+	lock    sync.RWMutex
+	backend ImageProxyBackend
+}
+
+// An ImageProxyBackend provides the functionality for different types of image proxies. An ImageProxy will construct
+// the required backend depending on the ImageProxySettings provided by the ConfigService.
+type ImageProxyBackend interface {
+	// GetImage provides a proxied image in response to an HTTP request.
+	GetImage(w http.ResponseWriter, r *http.Request, imageURL string)
+
+	// GetProxiedImageURL returns the URL to access a given image through the image proxy, whether the image proxy is
+	// running externally or as part of the Mattermost server itself.
+	GetProxiedImageURL(imageURL string) string
+
+	// GetUnproxiedImageURL returns the original URL of an image from one that has been directed at the image proxy.
+	GetUnproxiedImageURL(proxiedURL string) string
+}
+
+func MakeImageProxy(configService configservice.ConfigService, httpService httpservice.HTTPService) *ImageProxy {
+	proxy := &ImageProxy{
+		ConfigService: configService,
+		HTTPService:   httpService,
+	}
+
+	proxy.lock.Lock()
+	defer proxy.lock.Unlock()
+
+	proxy.configListenerId = proxy.ConfigService.AddConfigListener(proxy.OnConfigChange)
+
+	config := proxy.ConfigService.Config()
+	proxy.backend = proxy.makeBackend(*config.ImageProxySettings.Enable, *config.ImageProxySettings.ImageProxyType)
+
+	return proxy
+}
+
+func (proxy *ImageProxy) makeBackend(enable bool, proxyType string) ImageProxyBackend {
+	if !enable {
+		return nil
+	}
+
+	switch proxyType {
+	case model.IMAGE_PROXY_TYPE_LOCAL:
+		return makeLocalBackend(proxy)
+	case model.IMAGE_PROXY_TYPE_ATMOS_CAMO:
+		return makeAtmosCamoBackend(proxy)
+	default:
+		return nil
+	}
+}
+
+func (proxy *ImageProxy) Close() {
+	proxy.lock.Lock()
+	defer proxy.lock.Unlock()
+
+	proxy.ConfigService.RemoveConfigListener(proxy.configListenerId)
+}
+
+func (proxy *ImageProxy) OnConfigChange(oldConfig, newConfig *model.Config) {
+	if *oldConfig.ImageProxySettings.Enable != *newConfig.ImageProxySettings.Enable ||
+		*oldConfig.ImageProxySettings.ImageProxyType != *newConfig.ImageProxySettings.ImageProxyType {
+		proxy.lock.Lock()
+		defer proxy.lock.Unlock()
+
+		proxy.backend = proxy.makeBackend(*newConfig.ImageProxySettings.Enable, *newConfig.ImageProxySettings.ImageProxyType)
+	}
+}
+
+// GetImage takes an HTTP request for an image and requests that image using the image proxy.
+func (proxy *ImageProxy) GetImage(w http.ResponseWriter, r *http.Request, imageURL string) {
+	proxy.lock.RLock()
+	defer proxy.lock.RUnlock()
+
```

Comment:
I wanted to be careful that the proxy doesn't change its settings while handling a request. I'm not 100% sure of all the cases, but at the very least, if someone were to disable the image proxy between the below if statement (line 92) and the call to the backend (line 97), `proxy.backend` would be nil so it would error out.

I'll take a few minutes to see if I can come up with a good way to do this without a lock here since that would remove a bit of performance overhead caused by the lock. Handling the above case would actually be easy since we could just do
```go
backend := proxy.backend
// do stuff here since backend won't change on us
```
to get rid of the lock for these methods, although I'm not sure if there's anything that break the backend if it changed while they're running (like say the config settings change part way through fetching an image)

Reply:
I think I'm going to leave this as-is for now since I don't understand all of the changes required to remove the mutex on the `ImageProxy`, and I don't think the overhead from it is enough to be a concern at this time since the network request made by this method will be far more expensive.